### PR TITLE
feat(insights): automatically add insights middleware

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

Automatically add the insights middleware

**Result**

[FX-2248](https://algolia.atlassian.net/browse/FX-2248)

- add the insights middleware automatically
- identify internal vs. user-provided middleware
- make sure adding or removing the automatic insights middleware doesn't cause a search
- when a user adds a different insights middleware, it takes precedence
- update a few tests that now have `clickAnalytics: true`

[FX-2248]: https://algolia.atlassian.net/browse/FX-2248?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by aymeric-giraudet — APPROVED** [review]

**Comment by aymeric-giraudet** [comment]
Nice ! 🙌

File: packages/instantsearch.js/src/lib/InstantSearch.ts [hunk 1]
```diff
@@ -297,12 +298,17 @@ See ${createDocumentationLink({
...
     if (routing) {
-      const routerOptions = typeof routing === 'boolean' ? undefined : routing;
+      const routerOptions = typeof routing === 'boolean' ? {} : routing;
+      routerOptions.$$internal = true;
       this.use(createRouterMiddleware(routerOptions));
     }
 
+    // This is the default middleware,
+    // any user-provided middleware will be added later and override this one.
+    this.use(createInsightsMiddleware({ $$internal: true }));
```
Comment (Haroenv) [inline packages/instantsearch.js/src/lib/InstantSearch.ts:308]: something still needs to be fixed here, the script is always loaded from the internal middleware, even if users add a different middleware later (even right after). Not too sure how to fix that yet

Comment (Haroenv) [reply packages/instantsearch.js/src/lib/InstantSearch.ts:308]: solved through https://github.com/algolia/instantsearch/pull/5493

File: packages/instantsearch.js/src/connectors/infinite-hits/connectInfiniteHits.ts [hunk 2]
```diff
@@ -153,6 +153,11 @@ function getStateWithoutPage(state: PlainSearchParameters) {
   return rest;
 }
 
+function normalizeState(state: PlainSearchParameters) {
+  const { clickAnalytics, userToken, ...rest } = state || {};
+  return rest;
+}
```
Comment (Haroenv) [inline packages/instantsearch.js/src/connectors/infinite-hits/connectInfiniteHits.ts:159]: if infiniteHits is added before start, it's reading from the cache before the insights middleware is started, and thus it doesn't have clickAnalytics + userToken yet

Comment (Haroenv) [reply packages/instantsearch.js/src/connectors/infinite-hits/connectInfiniteHits.ts:159]: to solve this, we could change some things about the insights middleware to make sure we have the search parameters right away, even before init, but that's currently not possible yet (helper is null)

**Pull request merged**